go.mod

```diff
@@ -70,6 +70,7 @@ require (
 	github.com/alibabacloud-go/tea v1.2.0
 	github.com/alibabacloud-go/tea-utils/v2 v2.0.2
 	github.com/aliyun/credentials-go v1.3.0
+	github.com/artashesbalabekyan/barbican-sdk-go v1.2.0
```

Comment:
Hey, why did you create your own sdk, is there any reason to not use https://github.com/gophercloud/gophercloud ?

The barbican API doesn't allow you to read a secret value by its name, instead it only uses UUIDs, which is quite unhandy to work with. 

I see you return the first value from the list of matched keys [1], does it always returns the exact match as first item or are they alphabetically sorted? e.g. would `_foo` or `afoo` be returned before `foo` when we search for `foo` ?

[1] https://github.com/artashesbalabekyan/barbican-sdk-go/blob/main/client/secrets.go#L61-L65

Reply:
Hi.
I just couldn't find the gophercloud in the official docs. Didn't know that there is SDK in Golang.
https://www.openstack.org/software/project-navigator/sdks

Reply:
And about the names. It is possible to use the barbican with UUID, so the user can specify the UUID of the key in the ExternalSecret resource. But I have to think about how to do the same with PushSecret.

Reply:
UUIDs are not so easy to handle from a user's perspective. Looks like this is a limitation of the barbican API, that's very unfortunate. We would have to list + pattern match to provide a good UX :unamused: Not pretty but OK :smile: 

Out of curiosity: how do you plan to manage secrets via barbican? Create them via CLI and pull the via ESO using their UUID?

Reply:
Barbican supports the metadata. So we can use the metadata to mark the keys when we Push the key. But how to get the key by its name? It is problematic. So the behavior can be unpredictable. If I create multiple keys with the same name in the same second, it will be impossible to distinguish which is the last.

Reply:
What do you mean (list + pattern match)? Which parameters and criteria use for it?

Reply:
I think we can do it this way
```
data:
  - secretKey: secretValue
    remoteRef:
      key:<key_name>      # name of the Barbican secret key
      property: <key_uuid>  #uuid of the Barbican secret key
```

It is not good from a UX perspective. But if the user wants or has to use Barbican, he has to do it this way

Reply:
> What do you mean (list + pattern match)

We can list secrets and filter by name with `http://localhost:9311/v1/secrets?name=foo`.
That returns a list of all secrets matching the `name` exactly (it doesn't pattern match as i expected it would do). As you correctly pointed out, there's the possibility that a user creates two secrets with the same name. 
I'm not sure what the best approach is. I see the following options:
1. always return the first entry
2. return an error if more than one entry is found.
3. do NOT accept secrets by `name`, instead force user to work with UUIDs
 
I'm leaning towards 2, though 3 would also be acceptable. We shouldn't aim for option 1, IMO.

---

Your suggested use of `property` is contrary to the implementation in all other providers. The `property` should be used to extract a nested value from a json document.


Regarding managing different types for the key lookup i suggest this, though i'm open for alternative approaches. 
This is a pattern that has been implemented in AWS Secrets Manager.

```yaml
data:
  - secretKey: secretValue
    remoteRef:
      # example 1: lookup secret by name
      key: my-password 
      # example 2: lookup by uuid
      key: uuid/136316f9-d7e3-4118-8afc-ab0ad80bbf4d 
```

Reply:
I recommend variant 1 but using also sorting by the date.
`http://localhost:9311/v1/secrets?name=foo&sort=created:desc`
It will return the last entry with that name.
The only case it will not work is if the user creates multiple keys with the same name in the exact second. But the probability of this is tiny. We can warn about it in the documentation. What do you think about this approach?

Reply:
And also we can support the approach with uuid.
Like `key: uuid/136316f9-d7e3-4118-8afc-ab0ad80bbf4d `
But if the user wants to use it with names we should warn about the case above in the documentation.

Reply:
And also we can return an error only if there are multiple keys with the same datetime. Which is a very rare event.

Reply:
I pushed the changes. You can see and review it.